### DISCUSSION TITLE
EFI 1.1 Compatibility of Print function

### DISCUSSION THREAD

**Discussion by 0nelight** [ncroxon/gnu-efi#43] [Q&A] [answered]
Hi everyone,

I come from a Python background and I am a very beginner in C, EFI and gnu-efi but I managed to get a working utility for disabling defective ram on my MacBookPro Late 2013 using "Allocate Pages".

[https://github.com/0nelight/macOS-Disable-RAM-Areas](https://github.com/0nelight/macOS-Disable-RAM-Areas)

My problem is that I want to share this solution, and while it works for one user, another user reports of hangs of the program.

I relied heavily on ChatGPT for my previous (working for me and at least one other user) code, but now I want to do it by hand to find out why the hang is occuring for some users.

Now I thought I start with something very simple - writing "Hello World" with the functions of the gnu-efi program to abstract away the stuff I did before to make the code more clean.

But when using the print function, nothing gets printed when executing this via rEFInd Boot Manager.

```
#include <efi.h>
#include <efilib.h>

EFI_STATUS
efi_main (EFI_HANDLE ImageHandle, EFI_SYSTEM_TABLE *SystemTable) {
    InitializeLib(ImageHandle, SystemTable);

    Print(L"Hello World!\n");
    
    uefi_call_wrapper(BS->Stall, 1, 3000000);

    return EFI_SUCCESS;
}
```

Also using the uefi-call-wrapper for calling the conOut doesn't bring me success. Strangly the Stall command beneath it (which gets also called with the uefi-wrapper) works.

```
#include <efi.h>
#include <efilib.h>

#define GNU_EFI_3_0_COMPAT

EFI_STATUS
EFIAPI
efi_main (EFI_HANDLE ImageHandle, EFI_SYSTEM_TABLE *SystemTable) {
    InitializeLib(ImageHandle, SystemTable);

    UINTN index;
    
    uefi_call_wrapper(SystemTable->ConOut->OutputString, 2, SystemTable->ConOut, u"Hello application started\r\n");
    uefi_call_wrapper(SystemTable->ConOut->OutputString, 2, SystemTable->ConOut, u"\r\n\r\n\r\nHit any key to exit\r\n");
    uefi_call_wrapper(SystemTable->BootServices->WaitForEvent, 3, 1, &SystemTable->ConIn->WaitForKey, &index);

    uefi_call_wrapper(BS->Stall, 1, 3000000);

    return EFI_SUCCESS;
}
```

Adding EFIAPI or not before efi_main doesn't make a difference regarding the print function or the uefi_call_wrapper with ConOut.

How can I get to the bottom of why print doesn't work? For my compiler settings they are [here](https://github.com/0nelight/macOS-Disable-RAM-Areas/blob/main/Makefile) using the [5560e570](https://github.com/ncroxon/gnu-efi/tree/5560e57038e1579ed0b794d3edeb48ab6571ba86) commit of you here on GitHub.

Thank You
😊
Christian

**Comment by gmbr3** [comment, marked as answer]
Looks like objcopy is missing .rodata?

**Comment by 0nelight** [comment]
Thank you so much! It is working now with rodata inserted! :100:
